engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala

```diff
@@ -0,0 +1,273 @@
+  * This actor does not need nor care about cache hit/miss.
+  */
+class CallCacheHashingJobActor(jobDescriptor: BackendJobDescriptor,
+                               callCacheReadingJobActor: Option[ActorRef],
+                               initializationData: Option[BackendInitializationData],
+                               runtimeAttributeDefinitions: Set[RuntimeAttributeDefinition],
+                               backendName: String,
+                               fileHashingActorProps: Props,
+                               writeToCache: Boolean
+                              ) extends LoggingFSM[CallCacheHashingJobActorState, CallCacheHashingJobActorData] {
+
+  val fileHashingActor = makeFileHashingActor()
+
+  // Watch the read actor, as it will die when it's done (cache miss or successful cache hit)
+  // When that happens we want to either stop if writeToCache is false, or keep going
+  callCacheReadingJobActor map context.watch
+
+  initializeCCHJA()
+
+  override def preStart(): Unit = {
+    if (callCacheReadingJobActor.isEmpty && !writeToCache) {
+      log.error("Programmer error ! There is no reason to have a hashing actor if both read and write to cache are off")
+      context stop self
+    }
+    super.preStart()
+  }
+
+  when(WaitingForHashFileRequest) {
+    case Event(NextBatchOfFileHashesRequest, data) =>
+      data.fileHashRequestsRemaining.headOption match {
+        case Some(batch) if batch.nonEmpty =>
+          batch foreach { fileHashingActor ! _ }
+          goto(HashingFiles)
+        case _ =>
+          sendToCCReadActor(NoFileHashesResult, data)
+          stopAndStay(Option(NoFileHashesResult))
+      }
+  }
+
+  when(HashingFiles) {
+    case Event(FileHashResponse(result), data) =>
+      addFileHash(result, data) match {
+        case (newData, Some(partialHashes: PartialFileHashingResult)) =>
+          sendToCCReadActor(partialHashes, data)
+          // If there is no CCReader, send a message to itself to trigger the next batch
+          if (newData.callCacheReadingJobActor.isEmpty) self ! NextBatchOfFileHashesRequest
+          goto(WaitingForHashFileRequest) using newData
+        case (newData, Some(finalResult: FinalFileHashingResult)) =>
+          sendToCCReadActor(finalResult, newData)
+          stopAndStay(Option(finalResult))
+        case (newData, None) =>
+          stay() using newData
+      }
+  }
+
+  whenUnhandled {
+    case Event(Terminated(actor), data) if !writeToCache =>
+      stopAndStay(None)
+    case Event(Terminated(actor), data) if writeToCache =>
+      if (stateName == WaitingForHashFileRequest) self ! NextBatchOfFileHashesRequest
+      stay() using data.copy(callCacheReadingJobActor = None)
+    case Event(error: HashingFailedMessage, data) =>
+      log.error(error.reason, s"Failed to hash ${error.file}")
+      sendToCCReadActor(error, data)
+      context.parent ! error
+      stopAndStay(None)
+  }
```

Comment:
FWIW I've never really liked our penchant for specially handling the unhandled case unless we were going to do something other than log. As you said, akka already does this.